### PR TITLE
Improve thunk documentation

### PR DESCRIPTION
**Pull request by srghma**

### CODE REVIEW

**Comment by mightyiam** [comment]
Hey, bro. I am not familiar with the thunk module. I have a suggestion. If you have some free time, perhaps we could have a screen sharing session in which you'd explain the thunk module to me and I'd explain to you how to squash multiple commits into one?

**Comment by srghma** [comment]
>  you have some free time, perhaps we could have a screen sharing session in which you'd explain the thunk module to me

the session is possible, you can write me directly to srghma@gmail.com, but I'm better off keeping everything in written form (because it may be useful for others), unless you have something else in mind

> I'd explain to you how to squash multiple commits into one 

I know how to squash, tnx, will do

_____

I'm not the creator of this lib, can only tell what I've read :smile:

so

1. `fn` and `args` are stored in `VdomData`

https://github.com/snabbdom/snabbdom/blob/0fe8fd323770012b224253b8857edaa37f77dbe2/src/vnode.ts#L34-L35

2. 

the `vnode` (old) in `copyToThunk`  

https://github.com/snabbdom/snabbdom/blob/0fe8fd323770012b224253b8857edaa37f77dbe2/src/thunk.ts#L18

is `vnode` argument from [example](https://github.com/snabbdom/snabbdom#example) 

```
// Second `patch` invocation
patch(vnode, newVnode); // Snabbdom efficiently updates the old view to the new state
```

the `thunk` is `newVnode`

3. it's important to understand that old `vnode` will not be used anywhere any more, so we can mutate it as we want if it is faster, as here 

_____

IF thunk was fast-compared (`fn` and `args` are the same)

THEN

https://github.com/snabbdom/snabbdom/blob/0fe8fd323770012b224253b8857edaa37f77dbe2/src/snabbdom.ts#L282

no

https://github.com/snabbdom/snabbdom/blob/0fe8fd323770012b224253b8857edaa37f77dbe2/src/snabbdom.ts#L287

maybe 

https://github.com/snabbdom/snabbdom/blob/0fe8fd323770012b224253b8857edaa37f77dbe2/src/snabbdom.ts#L289

they are equal, no update

https://github.com/snabbdom/snabbdom/blob/0fe8fd323770012b224253b8857edaa37f77dbe2/src/snabbdom.ts#L298

they are equal, no update

File: README.md [hunk 1]
```diff
@@ -203,7 +203,7 @@ var vnode = h('div', {style: {color: '#000'}}, [
 
 ### `snabbdom/tovnode`
 
-Converts a DOM node into a virtual node. Especially good for patching over an pre-existing, 
+Converts a DOM node into a virtual node. Especially good for patching over an pre-existing,
```
Comment (mightyiam) [inline README.md:206]:
```suggestion
Converts a DOM node into a virtual node. Especially good for patching over an pre-existing, 
```

File: README.md [hunk 2]
```diff
@@ -740,12 +742,12 @@ Here are some approaches to building applications with Snabbdom.
   A JavaScript library for rendering html. Tung helps to divide html and JavaScript development.
 * [sprotty](https://github.com/theia-ide/sprotty) - "A web-based diagramming framework" uses Snabbdom.
 * [Mark Text](https://github.com/marktext/marktext) - "Realtime preview Markdown Editor" build on Snabbdom.
-* [puddles](https://github.com/flintinatux/puddles) - 
+* [puddles](https://github.com/flintinatux/puddles) -
```
Comment (mightyiam) [inline README.md:745]:
```suggestion
* [puddles](https://github.com/flintinatux/puddles) - 
```

File: README.md [hunk 3]
```diff
@@ -740,12 +742,12 @@ Here are some approaches to building applications with Snabbdom.
...
   "Tiny vdom app framework. Pure Redux. No boilerplate." - Built with :heart: on Snabbdom.
 * [Backbone.VDOMView](https://github.com/jcbrand/backbone.vdomview) - A [Backbone](http://backbonejs.org/) View with VirtualDOM capability via Snabbdom.
 * [Rosmaro Snabbdom starter](https://github.com/lukaszmakuch/rosmaro-snabbdom-starter) - Building user interfaces with state machines and Snabbdom.
 * [Pureact](https://github.com/irony/pureact) - "65 lines implementation of React incl Redux and hooks with only one dependency - Snabbdom"
-* [Snabberb](https://github.com/tobymao/snabberb) - A minimalistic Ruby framework using [Opal](https://github.com/opal/opal) and Snabbdom for building reactive views.  
+* [Snabberb](https://github.com/tobymao/snabberb) - A minimalistic Ruby framework using [Opal](https://github.com/opal/opal) and Snabbdom for building reactive views.
```
Comment (mightyiam) [inline README.md:750]:
```suggestion
* [Snabberb](https://github.com/tobymao/snabberb) - A minimalistic Ruby framework using [Opal](https://github.com/opal/opal) and Snabbdom for building reactive views.  
```

File: src/thunk.ts [hunk 4]
```diff
@@ -16,6 +16,7 @@ export interface ThunkFn {
 }
 
 function copyToThunk (vnode: VNode, thunk: VNode): void {
+  // mutate vnode.data because it's faster than using Object.assign later
```
Comment (mightyiam) [inline src/thunk.ts:19]: Which `Object.assign` is this referring to, please?

File: README.md [hunk 5]
```diff
@@ -583,6 +583,8 @@ arguments.
 
 `thunk(selector, key, renderFn, [stateArguments])`
 
+The `renderFn` is envoked if and only if the `renderFn` is changed or `[stateArguments]` array length or it's elements is changed.
```
Comment (mightyiam) [inline README.md:586]:
```suggestion
The `renderFn` is envoked only if the `renderFn` is changed or `[stateArguments]` array length or it's elements are changed.
```

File: src/thunk.ts (hunk 4 above)
Comment (srghma) [reply src/thunk.ts:19]: this one https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

File: README.md (hunk 1 above)
Comment (srghma) [reply README.md:206]: these whitespaces were removed by my editor automatically. I think this is good to not leave trailing whitespaces

Comment (mightyiam) [reply README.md:206]: It's good. Yet, I am strict about commit scope.

File: src/thunk.ts (hunk 4 above)
Comment (mightyiam) [inline src/thunk.ts:19]:
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -583,6 +583,8 @@ arguments.
 
 `thunk(selector, key, renderFn, [stateArguments])`
 
+The `renderFn` is envoked only if the `renderFn` is changed or `[stateArguments]` array length or it's elements are changed.
```
Comment (mightyiam) [inline README.md:586]:
```suggestion
The `renderFn` is invoked only if the `renderFn` is changed or `[stateArguments]` array length or it's elements are changed.
```

**Comment by mightyiam** [comment]
@srghma the code comment seemed like a different issue. Please feel free to add comments to the code of any module. Especially where it seems not clear.

**Comment by srghma** [comment]
thank you for merging it!)

**Pull request merged**